src/Interpreters/Aggregator.cpp

```diff
@@ -2631,45 +2685,75 @@ void NO_INLINE Aggregator::mergeStreamsImplCase(
     {
         /// Merge state of aggregate functions.
         aggregate_functions[j]->mergeBatch(
-            0, rows,
+            row_begin, row_end,
             places.get(), offsets_of_aggregate_states[j],
-            aggregate_columns[j]->data(),
+            aggregate_columns_data[j]->data(),
             aggregates_pool);
     }
+}
+
+template <typename Method, typename Table>
+void NO_INLINE Aggregator::mergeStreamsImpl(
+    Block & block,
+    Arena * aggregates_pool,
+    Method & method,
+    Table & data,
+    AggregateDataPtr overflow_row,
+    bool no_more_keys) const
+{
+    const AggregateColumnsConstData & aggregate_columns_data = params.makeAggregateColumnsData(block);
+    const ColumnRawPtrs & key_columns = params.makeRawKeyColumns(block);
+
+    mergeStreamsImpl<Method, Table>(
+        aggregates_pool,
+        method,
+        data,
+        overflow_row,
+        no_more_keys,
+        0,
+        block.rows(),
+        aggregate_columns_data,
+        key_columns);
 
     /// Early release memory.
     block.clear();
 }
 
 template <typename Method, typename Table>
 void NO_INLINE Aggregator::mergeStreamsImpl(
-    Block & block,
     Arena * aggregates_pool,
     Method & method,
     Table & data,
     AggregateDataPtr overflow_row,
-    bool no_more_keys) const
+    bool no_more_keys,
+    size_t row_begin,
+    size_t row_end,
+    const AggregateColumnsConstData & aggregate_columns_data,
+    const ColumnRawPtrs & key_columns) const
 {
     if (!no_more_keys)
-        mergeStreamsImplCase<false>(block, aggregates_pool, method, data, overflow_row);
+        mergeStreamsImplCase<false>(aggregates_pool, method, data, overflow_row, row_begin, row_end, aggregate_columns_data, key_columns);
     else
-        mergeStreamsImplCase<true>(block, aggregates_pool, method, data, overflow_row);
+        mergeStreamsImplCase<true>(aggregates_pool, method, data, overflow_row, row_begin, row_end, aggregate_columns_data, key_columns);
 }
 
 
-void NO_INLINE Aggregator::mergeWithoutKeyStreamsImpl(
+void NO_INLINE Aggregator::mergeBlockWithoutKeyStreamsImpl(
     Block & block,
     AggregatedDataVariants & result) const
 {
-    AggregateColumnsConstData aggregate_columns(params.aggregates_size);
-
-    /// Remember the columns we will work with
-    for (size_t i = 0; i < params.aggregates_size; ++i)
-    {
-        const auto & aggregate_column_name = params.aggregates[i].column_name;
-        aggregate_columns[i] = &typeid_cast<const ColumnAggregateFunction &>(*block.getByName(aggregate_column_name).column).getData();
-    }
+    AggregateColumnsConstData aggregate_columns = params.makeAggregateColumnsData(block);
+    mergeWithoutKeyStreamsImpl(result, 0, block.rows(), aggregate_columns);
 
+    /// Early release memory.
+    block.clear();
```

Comment:
If we always want to release memory here, it's better to change the signature to accept `Block` by value and use `std::move` to pass it. This explicitly express resource consumption.

Reply:
Good idea, done.